website/src/arcade/ArcadeEditor.tsx

```diff
@@ -0,0 +1,209 @@
+import * as React from "react";
+import * as monaco from "monaco-editor";
+import debounce from "lodash.debounce";
+import { MODELS } from "@site/src/arcade/models";
+import types from "@site/src/types.json";
+import * as q from "groqd";
+import { ArcadeDispatch, State } from "@site/src/arcade/state";
+import { evaluate, parse } from "groq-js";
+
+export type ArcadeEditorProps = {
+  dispatch: ArcadeDispatch;
+  query: State["query"];
+};
+
+export type ArcadeEditorHandle = {
+  setModel(model: keyof typeof MODELS): void;
+  runQuery(): void;
+};
+
+export const ArcadeEditor = React.forwardRef(
+  (
+    { dispatch, query }: ArcadeEditorProps,
+    ref: React.Ref<ArcadeEditorHandle>
+  ) => {
+    const containerRef = React.useRef<HTMLDivElement>(null);
+    const editorRef = React.useRef<monaco.editor.IStandaloneCodeEditor>();
+    const activeModel = "ts";
+
+    /**
+     * Execute TS query code, generates a query to store in state
+     */
+    const runCode = React.useCallback(async () => {
+      const editor = editorRef.current;
+      if (!editor) return;
+
+      try {
+        const model = MODELS.ts;
+        const worker = await monaco.languages.typescript.getTypeScriptWorker();
+        const client = await worker(model.uri);
+        const emitResult = await client.getEmitOutput(model.uri.toString());
+        const code = emitResult.outputFiles[0].text;
+
+        let playgroundRunQueryCount = 0;
+        const libs = {
+          groqd: q,
+          playground: {
+            runQuery: (
+              query: q.BaseQuery<any>,
+              params?: Record<string, string | number>
+            ) => {
+              playgroundRunQueryCount++;
+              if (playgroundRunQueryCount > 1) return;
+
+              try {
+                if (query instanceof q.BaseQuery) {
+                  dispatch({
+                    type: "INPUT_EVAL_SUCCESS",
+                    payload: { query, params },
+                  });
+                }
+              } catch {}
+            },
+          },
+        };
+        const scope = {
+          exports: {},
+          require: (name: keyof typeof libs) => libs[name],
+        };
+        const keys = Object.keys(scope);
+        new Function(...keys, code)(
+          ...keys.map((key) => scope[key as keyof typeof scope])
+        );
+      } catch {}
+    }, []);
+
+    /**
+     * Set up editor on mount
+     */
+    React.useEffect(() => {
+      const container = containerRef.current;
+      const editor = editorRef.current;
+
+      if (!container || editor) return;
+
+      monaco.languages.typescript.typescriptDefaults.setCompilerOptions({
+        typeRoots: ["groqd", "zod"],
+        target: monaco.languages.typescript.ScriptTarget.ES5,
+        strict: true,
+        skipLibCheck: true,
+        esModuleInterop: true,
+      });
+
+      monaco.languages.typescript.typescriptDefaults.setExtraLibs(extraLibs);
+
+      const handleContentChange = debounce(runCode, 500);
```

Comment:
Super minor, but you might want to `.cancel()` this in the cleanup, but the odds of this component dismounting mid schedule in 500ms is rare. But it could create a memory leak.